### PR TITLE
feat: support sni based tls route

### PR DESCRIPTION
**Pull request by mangoGoForward**
Signed-off-by: mango <xu.weiKyrie@foxmail.com>

### Type of change:

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### What this PR does / why we need it:

Please see #547

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

**Comment by tokers** [comment]
@mangoGoForward Please make the CI pass, thanks! See https://github.com/apache/apisix-ingress-controller/runs/6623221714?check_suite_focus=true.

**Review by tokers — COMMENTED** [review]
Please also add some e2e test cases to cover the SNI route feature.

**Comment by mangoGoForward** [comment]
> @mangoGoForward Please make the CI pass, thanks! See https://github.com/apache/apisix-ingress-controller/runs/6623221714?check_suite_focus=true.

I got some `connection refused` errors, this feature seems wouldn't effect the CI?

**Comment by mangoGoForward** [comment]
> Please also add some e2e test cases to cover the SNI route feature.

Thanks, I will do.

**Comment by mangoGoForward** [comment]
Hi @tokers How can I generate a `crd`? The `Makefile` seems has no related command.

**Comment by tao12345666333** [comment]
> n I generate a `crd`? Th

@mangoGoForward Unfortunately, current CRDs are manually maintained. We have https://github.com/apache/apisix-ingress-controller/issues/515 to track it.

**Comment by tao12345666333** [comment]
Please merge master latest code, and resolve conflicts. Thanks

File: samples/deploy/crd/v1/ApisixRoute.yaml [hunk 1]
```diff
@@ -134,7 +134,7 @@ spec:
                                       type: string
                                       minLength: 1
                                   required:
-                                  - scope
+                                    - scope
```
[the remark this answers is not in the document]
Comment (mangoGoForward) [reply samples/deploy/crd/v1/ApisixRoute.yaml:137]: It modified by `yamllint`, I think two indentations is more readable. there are many yaml files which uses no indentation and two indentations both, I suggest to support `yamllint` to  format those yaml files.

**Comment by mangoGoForward** [comment]
> Please merge master latest code, and resolve conflicts. Thanks

OK. I will solve it later.

File: samples/deploy/crd/v1/ApisixRoute.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tao12345666333) [reply samples/deploy/crd/v1/ApisixRoute.yaml:137]: > I suggest to support yamllint to format those yaml files.

Good idea. I will create new issue for track it.

[the remark this answers is not in the document]
Comment (tao12345666333) [reply samples/deploy/crd/v1/ApisixRoute.yaml:137]: https://github.com/apache/apisix-ingress-controller/issues/1062

[the remark this answers is not in the document]
Comment (tokers) [reply samples/deploy/crd/v1/ApisixRoute.yaml]: OK, so just ignore my comment here. Let's resolve it in another PR.

**Comment by tokers** [comment]
@mangoGoForward Please make the CI passed, it has some failures. Thanks!

**Comment by mangoGoForward** [comment]
> @mangoGoForward Please make the CI passed, it has some failures. Thanks!

OK,  I haven't found the the target of this error, it may takes a while to resolve.

**Comment by tao12345666333** [comment]
@mangoGoForward  hi, thanks for your contribution, we plan to enter v1.5 release window.

I have a few suggestions for this PR

* I would like to put this PR into v1.6
* Currently we have unified all resources to the v2 version of the API, I recommend implementing this feature only for the v2 version of ApisixRoute. Because when the v1.5 version is released, we will mark the v2beta3 version as deprecated.  You can check #707 for more detail

**Comment by tao12345666333** [comment]
Hi, do you have time to pick up this one? Thanks

**Comment by mangoGoForward** [comment]
> Hi, do you have time to pick up this one? Thanks

Yes, the file conflicts solved

**Comment by tao12345666333** [comment]
> 

@mangoGoForward Thanks for your active contribution!
As I said, we can implement this feature into the v2 API version.

File: samples/deploy/crd/v1/ApisixRoute.yaml [hunk 2]
```diff
@@ -521,6 +521,8 @@ spec:
                       match:
                         type: object
                         properties:
+                          host:
+                            type: string
```
Comment (tao12345666333) [inline samples/deploy/crd/v1/ApisixRoute.yaml:525]: please move this field to v2 api version section

File: pkg/kube/apisix/apis/config/v2beta3/types.go [hunk 3]
```diff
@@ -217,7 +217,8 @@ type ApisixRouteStream struct {
 type ApisixRouteStreamMatch struct {
 	// IngressPort represents the port listening on the Ingress proxy server.
 	// It should be pre-defined as APISIX doesn't support dynamic listening.
-	IngressPort int32 `json:"ingressPort" yaml:"ingressPort"`
+	IngressPort int32  `json:"ingressPort" yaml:"ingressPort"`
+	Host        string `json:"host,omitempty" yaml:"host,omitempty"`
```
Comment (tao12345666333) [inline pkg/kube/apisix/apis/config/v2beta3/types.go:221]: move this to  v2/types.go file.

**Review by tao12345666333 — CHANGES_REQUESTED** [review]

File: samples/deploy/crd/v1/ApisixRoute.yaml (hunk 2 above)
Comment (mangoGoForward) [reply samples/deploy/crd/v1/ApisixRoute.yaml]: Done

File: pkg/kube/apisix/apis/config/v2beta3/types.go (hunk 3 above)
Comment (mangoGoForward) [reply pkg/kube/apisix/apis/config/v2beta3/types.go]: Done

**Comment by tao12345666333** [comment]
LGTM !

could you please add e2e test cases to covet this one? thanks!

**Comment by mangoGoForward** [comment]
> LGTM !
> 
> could you please add e2e test cases to covet this one? thanks!

Sure, but have no time to deal with it recently, I will improve it as soon as possible

**Comment by tao12345666333** [comment]
Now that we're in the v1.6 development cycle, do you have time to  add e2e test cases to covet this one? Thanks!

**Comment by mangoGoForward** [comment]
> Now that we're in the v1.6 development cycle, do you have time to add e2e test cases to covet this one? Thanks!

Sorry reply to late, will be completed this week

**Comment by tao12345666333** [comment]
Thanks! @mangoGoForward

**Review by tao12345666333 — APPROVED** [review]
LGTM

But I think a more complete case can be added in e2e.
Create a certificate pair and use it for proxying.

**Review by gxthrj — APPROVED** [review]

**Comment by tao12345666333** [comment]
But we can do it in a follow-up PR  #1438  @mangoGoForward

**Comment by mangoGoForward** [comment]
> LGTM
> 
> But I think a more complete case can be added in e2e. Create a certificate pair and use it for proxying.

Thanks, it absolutely need more test cases to cover it. I need to study how the TLS route with SNI works and how to test it, I am so poor about knowledge in this area.

**Comment by tao12345666333** [comment]
Don't worry, you can continue if you are interested. Otherwise see if others are interested.

**Pull request merged**